File: frontend/src/codesearch.rs
/*
Copyright (C) 2020 Kunal Mehta <legoktm@member.fsf.org>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
/// codesearch querying "library" (nothing seed specific)
use anyhow::Result;
use reqwest::Client;
use serde::Deserialize;
use std::collections::{BTreeMap, HashMap};

#[derive(Deserialize)]
pub struct HoundResponse {
    #[serde(rename = "Error")]
    pub error: Option<String>,
    #[serde(rename = "Results")]
    pub results: Option<HoundResults>,
    /*#[serde(rename = "Stats")]
    stats: Option<HoundStats>,*/
}

pub type HoundResults = HashMap<String, RepoResult>;

#[derive(Deserialize, Clone)]
pub struct RepoResult {
    #[serde(rename = "Matches")]
    pub matches: Vec<RepoMatch>,
    #[serde(rename = "FilesWithMatch")]
    pub files_with_match: usize,
    #[serde(rename = "Revision")]
    pub revision: String,
}

#[derive(Deserialize, Clone)]
pub struct RepoMatch {
    #[serde(rename = "Filename")]
    pub(crate) filename: String,
    #[serde(rename = "Matches")]
    pub(crate) matches: Vec<Match>,
}

#[derive(Deserialize, Clone)]
pub struct Match {
    #[serde(rename = "Line")]
    line: String,
    #[serde(rename = "LineNumber")]
    line_number: u32,
    #[serde(rename = "Before")]
    before: Vec<String>,
    #[serde(rename = "After")]
    after: Vec<String>,
}

/*#[derive(Deserialize)]
struct HoundStats {
    #[serde(rename = "FilesOpened")]
    files_opened: u32,
    #[serde(rename = "Duration")]
    duration: u32,
}*/

pub fn backends() -> Vec<(String, String)> {
    vec![
        ("search".to_string(), "Everything".to_string()),
        ("core".to_string(), "MediaWiki core".to_string()),
        ("extensions".to_string(), "Extensions".to_string()),
        ("skins".to_string(), "Skins".to_string()),
        ("things".to_string(), "Extensions & things".to_string()),
        ("bundled".to_string(), "MW tarball".to_string()),
        ("deployed".to_string(), "Wikimedia deployed".to_string()),
        ("libraries".to_string(), "PHP libraries".to_string()),
        ("operations".to_string(), "Wikimedia Operations".to_string()),
        ("ooui".to_string(), "OOUI".to_string()),
        ("milkshake".to_string(), "Milkshake".to_string()),
        ("pywikibot".to_string(), "Pywikibot".to_string()),
        ("services".to_string(), "Wikimedia Services".to_string()),
        ("analytics".to_string(), "Analytics".to_string()),
    ]
}

pub fn is_valid_backend(profile: &str) -> bool {
    for (backend, _) in backends() {
        if &backend == profile {
            return true;
        }
    }

    false
}

#[derive(Clone, Debug)]
pub struct SearchOptions {
    pub query: String,
    pub files: String,
    pub repos: Option<String>,
    pub case_insensitive: bool,
}

// TODO: move to separate library
pub async fn send_query(
    options: &SearchOptions,
    profile: &str,
) -> Result<HoundResponse> {
    let opts = options.clone();
    let case_insensitive = if opts.case_insensitive {
        "fosho"
    } else {
        "nope"
    };
    // offset to query
    let rng = if opts.repos.is_some() { "20:" } else { ":20" };
    let resp = Client::new()
        .get(&format!(
            "https://codesearch.wmcloud.org/{}/api/v1/search",
            profile
        ))
        // stats=fosho&repos=*&rng=%3A20&q=class+LinkRenderer%5Cb&files=&i=nope
        .query(&[
            ("stats", "fosho"),
            ("repos", &opts.repos.unwrap_or("*".to_string())),
            ("rng", rng),
            ("q", &opts.query),
            ("files", &opts.files),
            ("i", case_insensitive),
        ])
        .send()
        .await?
        .json()
        .await?;
    Ok(resp)
}

pub type HoundConfig = HashMap<String, RepoConfig>;

#[derive(Deserialize)]
pub struct RepoConfig {
    pub url: String,
    #[serde(rename = "url-pattern")]
    pub url_pattern: UrlPattern,
}

#[derive(Deserialize)]
pub struct UrlPattern {
    #[serde(rename = "base-url")]
    pub base_url: String,
    pub anchor: String,
}

pub async fn fetch_config(profile: &str) -> Result<HoundConfig> {
    let resp = Client::new()
        .get(&format!(
            "https://codesearch.wmcloud.org/{}/api/v1/repos",
            profile
        ))
        .send()
        .await?
        .json()
        .await?;

    Ok(resp)
}

/// Represent matches in a more straightforward format for display
pub fn flatten(repomatch: &RepoMatch) -> File {
    let mut file = File::new(&repomatch.filename);
    for match_ in &repomatch.matches {
        // TODO: we need separators between non-consecutive lines
        file.lines.insert(
            match_.line_number,
            Line {
                highlight: true,
                text: match_.line.to_string(),
            },
        );
        // We hardcode that the length of before and after is 2
        // But if the before line already exists, we don't want to
        // overwrite highlight
        let before_length = match_.before.len();
        if before_length >= 1 {
            if !file.lines.contains_key(&(&match_.line_number - 2)) {
                file.lines.insert(
                    match_.line_number - 2,
                    Line {
                        highlight: false,
                        text: match_.before[0].to_string(),
                    },
                );
            }
        }

        if before_length == 2 {
            if !file.lines.contains_key(&(&match_.line_number - 1)) {
                file.lines.insert(
                    match_.line_number - 1,
                    Line {
                        highlight: false,
                        text: match_.before[1].to_string(),
                    },
                );
            }
        }
        let after_length = match_.after.len();
        if after_length >= 1 {
            file.lines.insert(
                match_.line_number + 1,
                Line {
                    highlight: false,
                    text: match_.after[0].to_string(),
                },
            );
        }

        if after_length == 2 {
            file.lines.insert(
                match_.line_number + 2,
                Line {
                    highlight: false,
                    text: match_.after[1].to_string(),
                },
            );
        }
    }

    file
}

pub struct File {
    pub name: String,
    // We use BTreeMap for its built-in ordering
    pub lines: BTreeMap<u32, Line>,
}

impl File {
    fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            lines: BTreeMap::new(),
        }
    }
}

pub struct Line {
    pub highlight: bool,
    pub text: String,
}

File: frontend/src/Codesearch.php
<?php
/**
 * Copyright (C) 2022 Timo Tijhof
 * Copyright (C) 2020 Kunal Mehta <legoktm@debian.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
namespace Wikimedia\Codesearch;

use RuntimeException;

class Codesearch {
	public const BACKEND_DEFAULT = 'search';
	public const BACKENDS = [
		'search' => 'Everything',
		'core' => 'MediaWiki core',
		'extensions' => 'Extensions',
		'skins' => 'Skins',
		'things' => 'MW extensions & skins',
		'bundled' => 'MW tarball',
		'libraries' => 'MW libraries',
		'deployed' => 'MediaWiki & services at WMF',
		'operations' => 'Wikimedia SRE',
		'puppet' => 'Puppet',
		'ooui' => 'OOUI',
		'milkshake' => 'Milkshake',
		'pywikibot' => 'Pywikibot',
		'services' => 'Wikimedia Services',
		'devtools' => 'CI & Development',
		'analytics' => 'Data Engineering',
		'wmcs' => 'Wikimedia Cloud Services',
		'armchairgm' => 'ArmchairGM',
		'shouthow' => 'ShoutHow',
		'apps' => 'Mobile Apps',
	];
	public const BACKENDS_HIDDEN = [
		'armchairgm' => true,
		'extensions' => true,
		'milkshake' => true,
		'ooui' => true,
		'services' => true,
		'shouthow' => true,
		'skins' => true,
	];
	public const HOUND_BASE_PUBLIC = 'https://codesearch-backend.wmcloud.org';

	private const USER_AGENT = 'codesearch-frontend <https://gerrit.wikimedia.org/g/labs/codesearch>';
	private const META_CACHE_TTL = 3600;

	private array $apcuStats = [
		'hits' => 0,
		'misses' => 0,
	];

	public function getBackendLabel( string $backend ): ?string {
		return self::BACKENDS[ $backend ] ?? null;
	}

	public function getApcuStats(): array {
		return $this->apcuStats + [ 'enabled' => function_exists( 'apcu_fetch' ) ];
	}

	protected function getWithSetCallback( string $key, int $ttl, $callback ) {
		$hasApcu = function_exists( 'apcu_fetch' );
		$val = $hasApcu ? apcu_fetch( $key ) : false;
		if ( $val !== false ) {
			$this->apcuStats['hits']++;
			$this->debug( sprintf( 'APCu cache hit for %s', $key ) );
		} else {
			$this->debug( sprintf( 'APCu cache miss for %s', $key ) );
			$val = $callback();
			if ( $hasApcu ) {
				apcu_store( $key, $val, $ttl );
				$this->apcuStats['misses']++;
			}
		}
		return $val;
	}

	/**
	 * @see CodesearchDebug::debug
	 * @param string $msg
	 */
	public function debug( $msg ): void {
	}

	private function getHoundApi( string $backend, $houndBase = null ): string {
		$houndBase ??= rtrim( getenv( 'CODESEARCH_HOUND_BASE' ) ?: self::HOUND_BASE_PUBLIC, '/' );
		return "$houndBase/$backend/api";
	}

	public function getHealth(): array {
		$houndBase = rtrim( getenv( 'CODESEARCH_HOUND_BASE' ) ?: self::HOUND_BASE_PUBLIC, '/' );
		$result = $this->getHttp( "$houndBase/_health.json" );
		$health = json_decode( $result, true );
		if ( !is_array( $health ) ) {
			trigger_error( "Hound /_health.json returned " . substr( $result, 0, 1024 ) );
			throw new ApiUnavailable( "Hound /_health.json returned invalid data" );
		}
		return $health;
	}

	public function formatPublicSearchApi( string $backend, array $fields ): string {
		$params = [
			// List 'q' first, matching robots.txt
			'q' => $fields['query'],
			'i' => $fields['caseInsensitive'] ? 'fosho' : null,
			'files' => $fields['filePath'],
			'excludeFiles' => $fields['excludeFiles'],
			'repos' => $fields['repos'] ?: '*',
			'stats' => 'fosho',
			// Enable rng ("offset:limit") to limit results to "page 1".
			// "Load more" is handled in codesearch.js.
			'rng' => ':20',
		];
		return $this->getHoundApi( $backend, self::HOUND_BASE_PUBLIC ) . '/v1/search?' . http_build_query( $params );
	}

	public function getCachedConfig( string $backend ): array {
		return $this->getWithSetCallback(
			"codesearch-config-v1:$backend",
			self::META_CACHE_TTL,
			function () use ( $backend ) {
				$url = $this->getHoundApi( $backend ) . '/v1/repos';
				$val = json_decode( $this->getHttp( $url ), true );
				if ( !$val ) {
					throw new ApiUnavailable( 'Hound /v1/repos returned empty or invalid data' );
				}
				// Optimization: Strip out data not needed by client
				foreach ( $val as $repoId => &$repoConf ) {
					$repoConf = [
						'url' => $repoConf['url'],
						'url-pattern' => $repoConf['url-pattern'],
					];
				}
				return $val;
			}
		);
	}

	public function getCachedExcludes( string $backend ): array {
		return $this->getWithSetCallback(
			"codesearch-excludes-v1:$backend",
			self::META_CACHE_TTL,
			function () use ( $backend ) {
				$urls = [];
				$reposData = $this->getCachedConfig( $backend );
				foreach ( $reposData as $repo => $_ ) {
					$urls[$repo] = $this->getHoundApi( $backend ) . '/v1/excludes?' . http_build_query( [ 'repo' => $repo ] );
				}
				$results = $this->getHttpMulti( $urls );

				$val = [];
				foreach ( $results as $repo => $result ) {
					$files = json_decode( $result, true );
					if ( !is_array( $files ) ) {
						trigger_error( "Hound /v1/excludes for $repo returned " . substr( $result, 0, 1024 ) );
						throw new ApiUnavailable( "Hound /v1/excludes for $repo returned invalid data" );
					}
					$val[] = [
						'repo' => $repo,
						'excludesCount' => count( $files ),
						'files' => $files,
					];
				}
				return $val;
			}
		);
	}

	protected function getHttp( string $url, $timeout = 3 ): string {
		$curlOptions = [
			CURLOPT_TIMEOUT => $timeout,
			CURLOPT_MAXREDIRS => 2,
			CURLOPT_FOLLOWLOCATION => true,
			CURLOPT_USERAGENT => self::USER_AGENT,
			CURLOPT_HTTPHEADER => [],
			CURLOPT_RETURNTRANSFER => true,
		];
		$curlHandle = curl_init( $url );
		if ( !curl_setopt_array( $curlHandle, $curlOptions ) ) {
			throw new RuntimeException( 'Could not set curl options' );
		}
		$this->debug( sprintf( 'getHttp timeout=%d %s', $timeout, $url ) );
		$curlRes = curl_exec( $curlHandle );
		if ( curl_errno( $curlHandle ) == CURLE_OPERATION_TIMEOUTED ) {
			throw new ApiUnavailable( "Hound request timed out after $timeout seconds" );
		}
		if ( $curlRes === false ) {
			throw new ApiUnavailable( 'Hound request failed: ' . curl_error( $curlHandle ) );
		}
		curl_close( $curlHandle );
		return $curlRes;
	}

	protected function getHttpMulti( array $urls, $throttled = false ): array {
		if ( !$throttled ) {
			$this->debug( sprintf( 'getHttpMulti chunking %d requests', count( $urls ) ) );

			$results = [];
			$t = null;
			foreach ( array_chunk( $urls, 99, true ) as $chunk ) {

				// Avoid HTTP 429 from WMCS dynamicproxy (ratelimit of 100/s)
				// Note that only the sleep is conditional, the chunking is not.
				// We don't want to send 1000+ requests at once, as they will not
				// complete within the timeout.
				if ( !getenv( 'CODESEARCH_HOUND_BASE' ) ) {
					// usleep in microseconds, hrtime in nanoseconds
					$remainingUs = !$t ? 0 : ceil( 1e6 - ( ( hrtime( true ) - $t ) / 1000 ) );
					if ( $remainingUs > 0 ) {
						$this->debug( sprintf( 'getHttpMulti sleeping %dms between chunks', $remainingUs / 1000 ) );
						usleep( $remainingUs );
					}
					$t = hrtime( true );
				}

				$results += $this->getHttpMulti( $chunk, true );
			}
			return $results;
		}

		$cmh = curl_multi_init();
		curl_multi_setopt( $cmh, CURLMOPT_MAXCONNECTS, 50 );
		curl_multi_setopt( $cmh, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX );

		$infos = [];
		$curlOptions = [
			CURLOPT_TIMEOUT => 3,
			CURLOPT_MAXREDIRS => 2,
			CURLOPT_FOLLOWLOCATION => true,
			CURLOPT_USERAGENT => self::USER_AGENT,
			CURLOPT_HTTPHEADER => [],
			CURLOPT_RETURNTRANSFER => true,
			CURLOPT_HEADERFUNCTION => static function ( $ch, $header ) use ( &$infos ) {
				$len = strlen( $header );
				// HTTP/2 responds with only a number, no textual reason as well
				if ( preg_match( "/^HTTP\/(?:1\.[01]|2) (\d+)/", $header, $m ) ) {
					$infos[ (int)$ch ]['status'] = (int)$m[1];
				}
				return $len;
			},
			// Inspired by MultiHttpClient from MediaWiki 1.42
			CURLOPT_PIPEWAIT => 1,
		];

		$this->debug( sprintf( 'getHttpMulti with %d requests', count( $urls ) ) );
		foreach ( $urls as $urlKey => $url ) {
			$curlHandle = curl_init( $url );
			if ( !curl_setopt_array( $curlHandle, $curlOptions ) ) {
				throw new RuntimeException( 'Could not set curl options' );
			}
			$infos[(int)$curlHandle] = [
				'urlKey' => $urlKey,
				'handle' => $curlHandle,
				'result' => null,
				'status' => null,
			];
			curl_multi_add_handle( $cmh, $curlHandle );
		}

		// Execute multiple handles at once
		$active = null;
		do {
			$status = curl_multi_exec( $cmh, $active );
			if ( $active ) {
				curl_multi_select( $cmh );
			}
			// phpcs:ignore MediaWiki.ControlStructures.AssignmentInControlStructures
			while ( ( $info = curl_multi_info_read( $cmh ) ) !== false ) {
				/** @var $info false|array{msg:int,result:int,handle:object} */
				$infos[ (int)$info['handle'] ]['result'] = $info['result'];
			}
		} while ( $active > 0 && $status == CURLM_OK );

		'@phan-var array{urlKey:string,handle:object,result:?int,status:?int}[] $infos';

		$results = [];
		foreach ( $infos as $_ => $info ) {
			$urlKey = $info['urlKey'];
			if ( $info['result'] !== null && $info['result'] !== 0 ) {
				throw new ApiUnavailable( "Request {$urls[$urlKey]} failed " . curl_strerror( $info['result'] ) );
			}
			if ( $info['status'] >= 400 ) {
				throw new ApiUnavailable( "Request {$urls[$urlKey]} failed HTTP " . $info['status'] );
			}
			$result = curl_multi_getcontent( $info['handle'] );
			curl_multi_remove_handle( $cmh, $info['handle'] );
			curl_close( $info['handle'] );
			$results[$urlKey] = $result;
		}

		curl_multi_close( $cmh );
		return $results;
	}
}
